addons/project_timesheet_holidays/models/hr_holidays.py

```diff
@@ -56,45 +56,53 @@ class Holidays(models.Model):
     timesheet_ids = fields.One2many('account.analytic.line', 'holiday_id', string="Analytic Lines")
 
     def _validate_leave_request(self):
-        """ Timesheet will be generated on leave validation only if a timesheet_project_id and a
-            timesheet_task_id are set on the corresponding leave type. The generated timesheet will
-            be attached to this project/task.
+        """ Timesheet will be generated on leave validation only if timesheet_generate is True
+            If company is set, timesheet_project_id and timesheet_task_id from leave type are
+            used as project_id and task_id.
+            Else, internal_project_id and leave_timesheet_task_id are used.
+            The generated timesheet will be attached to this project/task.
         """
-        holidays = self.filtered(
-            lambda l: l.holiday_type == 'employee' and
-            l.holiday_status_id.timesheet_project_id and
-            l.holiday_status_id.timesheet_task_id and
-            l.holiday_status_id.timesheet_project_id.sudo().company_id == (l.holiday_status_id.company_id or self.env.company))
-
-        # Unlink previous timesheets do avoid doublon (shouldn't happen on the interface but meh)
-        old_timesheets = holidays.sudo().timesheet_ids
-        if old_timesheets:
-            old_timesheets.holiday_id = False
-            old_timesheets.unlink()
+        vals_list = []
+        leave_ids = []
+        for leave in self:
+            if leave.holiday_type != 'employee' or not leave.holiday_status_id.timesheet_generate:
+                continue
 
-        # create the timesheet on the vacation project
-        holidays._timesheet_create_lines()
+            if leave.holiday_status_id.company_id:
+                project_id, task_id = leave.holiday_status_id.timesheet_project_id, leave.holiday_status_id.timesheet_task_id
+            else:
+                project_id, task_id = leave.employee_id.company_id.internal_project_id, leave.employee_id.company_id.leave_timesheet_task_id
```

Comment:
I prefer naming `project` and `task` when those variables contain a recordset. If the variable will contain id then I name the variable `project_id`, I think it is clearer. 🤷🏼‍♂️ 
(just valid for variables naming, not the field)

```suggestion
            if leave.holiday_status_id.company_id:
                project, task = leave.holiday_status_id.timesheet_project_id, leave.holiday_status_id.timesheet_task_id
            else:
                project, task = leave.employee_id.company_id.internal_project_id, leave.employee_id.company_id.leave_timesheet_task_id
```